File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
import sys
input = sys.stdin.readline
def same(parent,x):
    if parent[x]!=x:
        parent[x] = same(parent,parent[x])
    return parent[x]
def union(parent, rank, a,b):
    root_a = same(parent, a)
    root_b = same(parent, b)
    if root_a != root_b :
        if rank[root_a]<rank[root_b]:
            parent[root_a] = root_b
        elif rank[root_a]>rank[root_b]:
            parent[root_b] = root_a
        else:
            parent[root_b] = root_a
            rank[root_a]+=1
n,m = map(int,input().split())
parent = [i for i in range(n+1)]
rank = [0] *(n+1)
res = []
for _ in range(m):
    op,a,b = map(int,input().split())
    if op==0: #합집합
        union(parent, rank,a,b)
    elif op==1: #같은지 확인
        if same(parent,a)==same(parent,b):
            res.append("YES")
        else:
            res.append("NO")
print("\n".join(res))

File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.*;
import java.util.*;

public class Main {
    static int find(int[] parents,int x){
        if(parents[x]!=x){
            parents[x] = find(parents,parents[x]);
        }return parents[x];
    }
    static void union(int[] parents,int[] rank, int a, int b){
        int rootA = find(parents, a);
        int rootB = find(parents, b);
        if(rootA !=rootB){
            if(rank[rootA]<rank[rootB]){
                parents[rootA] = rootB;
            }else if(rank[rootA]>rank[rootB]){
                parents[rootB]=rootA;
            }else{
                parents[rootB] = rootA;
                rank[rootA]++;
            }
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        String[] nm  = br.readLine().split(" ");
        int n = Integer.parseInt(nm[0]);
        int m = Integer.parseInt(nm[1]);

        int[] parents = new int[n+1];
        int[] rank = new int[n+1];
        for(int i=0;i<=n;i++){
            parents[i] = i;
            rank[i] = 0;
        }
        for(int i=0;i<m;i++){
            String[] input = br.readLine().split(" ");
            int op = Integer.parseInt(input[0]);
            int a = Integer.parseInt(input[1]);
            int b = Integer.parseInt(input[2]);
            if(op==0){
                union(parents, rank, a, b);
            }else if(op==1){
                if(find(parents,a)==find(parents, b)){
                    sb.append("YES\n");
                }else{
                    sb.append("NO\n");
                }
            }
        }
        System.out.print(sb);
    }
    }
    


